doc/developer/design/20220303_secrets.md

```diff
@@ -0,0 +1,214 @@
+section). Therefore there is presently no additional metadata that needs to be
+recorded about a secret.
+
+When serializing a secret, its `create_sql` will be synthesized as:
+
+```sql
+CREATE SECRET <name> AS '******'
+```
+
+When deserializing a secret on bootstrap, the `'******'` will be ignored.
+
+### Secrets controller
+
+The `Coordinator` will host a `SecretsController` that securely handles the
+storage of secrets. It will interact with the controller via the following
+interface.
+
+```rust
+/// Securely stores secrets.
+pub trait SecretsController {
+    /// Applies the specified secret operations in bulk.
+    ///
+    /// Implementations must apply the operations atomically. If the method
+    /// returns `Ok(())`, then all operations have been applied successfully;
+    /// if the method returns `Err(())`, then none of the operations have been
+    /// applied.
+    ///
+    /// Implementations are permitted to reject combinations of operations which
+    /// they cannot apply atomically.
+    fn apply(&mut self, ops: Vec<SecretOp>) -> Result<(), anyhow::Error>;
+}
+
+/// An operation on a [`SecretsController`].
+pub enum SecretOp {
+    /// Create or update the contents of a secret.
+    Ensure {
+        /// The ID of the secret to create or update.
+        id: GlobalId,
+        /// The binary contents of the secret.
+        contents: Vec<u8>,
+    },
+    /// Delete a secret.
+    Delete {
+        /// The id of the secret to delete.
+        id: GlobalId
+    }
+}
+```
+
+There will be two concrete implementations of the `SecretsController`:
+
+  * The `FilesystemSecretsController` will store secrets on the local filesystem.
+    It will not support multi-operation atomicity, but will support atomically
+    updating or removing one secret at a time.
+
+  * The `KubernetesSecretsController` will store secrets in Kubernetes. It
+    will support multi-operation atomicity.
+
+All secrets controllers expose the contents of secrets to the [`STORAGE`]
+and [`COMPUTE`] layer by way of a directory that contains one file per secret,
+where the name of the file is the secret's ID and the contents of the file are
+the secret's contents. For example, if IDs `u12` and `u36` are secrets, the
+directory would look like this:
+
+```
+secrets/
+  u12
+  u36
+```
+
+This bears repeating: even the `KubernetesSecretsController` will expose the
+contents of the secrets via the filesystem. The filesystem is the *lingua
+franca* for exposing secrets to applications. Kubernetes supports exposing the
+contents of secrets as files on the filesystem, as does [HashiCorp Vault].
+
+The [`STORAGE`] and [`COMPUTE`] layers will need to be taught to re-read the
+contents of secrets that they consume from disk periodically, in case the
+contents have bene updated by an `ALTER SECRET`. We may wish to build a helper
+"secret consumer" library that uses filesystem events to notice when the secrets
+change, rather than polling periodically.
```

Comment:
just wanna add a big 👍 here. funneling everything through the filesystem + periodically re-reading secrets from disk is operationally very nice to work with